File: src/medium/MinStack.java
package medium;

import java.util.Stack;

/**
 * Created by shirui on 10/23/16.
 */
public class MinStack {
  /**
   * initialize your data structure here.
   */
  Stack<Integer> minStack;
  Stack<Integer> stack;

  public MinStack() {
    minStack = new Stack<>();
    stack = new Stack<>();
  }

  public void push(int x) {
    stack.push(x);
    if (getMin() > x) {
      minStack.push(x);
    } else {
      minStack.push(getMin());
    }

  }

  public void pop() {
    minStack.pop();
    stack.pop();
  }

  public int top() {
    return stack.peek();
  }

  public int getMin() {
    return minStack.peek();
  }
}

File: csharp/LeetCode/DataStructures/MinStack.cs
namespace LeetCode.DataStructures
{
    using System.Collections.Generic;
    using System.Linq;

    public class MinStack
    {
        private readonly Stack<int> Values = new Stack<int>();
        private readonly Stack<int> Mins = new Stack<int>();
        /** initialize your data structure here. */
        public MinStack()
        {

        }

        public void Push(int val)
        {
            if (Mins.Any() && Mins.Peek() < val)
            {
                Mins.Push(Mins.Peek());
            }
            else
            {
                Mins.Push(val);
            }
            Values.Push(val);
        }

        public void Pop()
        {
            Mins.Pop();
            Values.Pop();
        }

        public int Top()
        {
            return Values.Peek();
        }

        public int GetMin()
        {
            return Mins.Peek();
        }
    }
}
